File: dataAnalysisCodes/watchERP_2stim/04-icon-detection/createSymbolCorrectnessDataset.m
function createSymbolCorrectnessDataset

% init host name
%--------------------------------------------------------------------------
if isunix,
    envVarName = 'HOSTNAME';
else
    envVarName = 'COMPUTERNAME';
end
hostName = lower( strtok( getenv( envVarName ), '.') );

switch hostName,
    case 'kuleuven-24b13c',
        addpath( genpath('d:\KULeuven\PhD\Work\Hybrid-BCI\HybBciCode\dataAnalysisCodes\deps\') );
        dataDir = 'd:\KULeuven\PhD\Work\Hybrid-BCI\HybBciRecordedData\watchERP_2stim\';
        resDir  = 'd:\KULeuven\PhD\Work\Hybrid-BCI\HybBciProcessedData\watchERP_2stim\';
        codeDir = 'd:\KULeuven\PhD\Work\Hybrid-BCI\HybBciCode\dataAnalysisCodes\watchERP_2stim\';
    case 'neu-wrk-0158',
        addpath( genpath('d:\Adrien\Work\Hybrid-BCI\HybBciCode\dataAnalysisCodes\deps\') );
        addpath( genpath('d:\Adrien\matlabToolboxes\eeglab10_0_1_0b\') );
        rmpath( genpath('d:\Adrien\matlabToolboxes\eeglab10_0_1_0b\external\SIFT_01_alpha') );
        dataDir = 'd:\Adrien\Work\Hybrid-BCI\HybBciRecordedData\watchERP_2stim\';
        resDir  = 'd:\Adrien\Work\Hybrid-BCI\HybBciProcessedData\watchERP_2stim\';
        codeDir = 'd:\Adrien\Work\Hybrid-BCI\HybBciCode\dataAnalysisCodes\watchERP_2stim\';
    case {'sunny', 'solaris', ''}
        addpath( genpath( '~/PhD/hybridBCI-stuffs/deps/' ) );
        rmpath( genpath('~/PhD/hybridBCI-stuffs/deps/eeglab10_0_1_0b/external/SIFT_01_alpha') );
        dataDir = '~/PhD/hybridBCI-stuffs/watchERP_2stim/data/';
        resDir  = '~/PhD/hybridBCI-stuffs/watchERP_2stim/results/';
        codeDir = '~/PhD/hybridBCI-stuffs/watchERP_2stim/code/';
    otherwise,
        error('host not recognized');
end

if isunix,
    TableName   = fullfile( codeDir, '01-preprocess-plot', filesep, 'watchErpDataset.csv');
    fileList    = dataset('File', TableName, 'Delimiter', ',');
else
    TableName   = fullfile( codeDir, '01-preprocess-plot', filesep, 'watchErpDataset.xlsx');
    fileList    = dataset('XLSFile', TableName);
end

run = unique( fileList.run );
if ~isequal(run(:), (1:max(run))'), error('wrong run numbering'); end
listRunsForTrain = {1 2 3, [1 2], [2 3], [3 4]};
% listTestRun      = cellfun(@(x) run( ~ismember(run, x)), listRunsForTrain, 'UniformOutput', false );
listTestRun      = cellfun(@(x) run( run > max(x) ), listRunsForTrain, 'UniformOutput', false );

%% ========================================================================================================

%--------------------------------------------------------------------------
sub                 = unique( fileList.subjectTag );
nSub                = numel(sub);
[~, folderName, ~]  = fileparts( fileparts( mfilename('fullpath') ) );
resDirOut           = fullfile( resDir, folderName );
if ~exist( resDirOut, 'dir' ), mkdir( resDirOut ); end


for ii = 1:numel( listRunsForTrain )
    
    fprintf('Treating case %d out of %d (train on runs %d\n', ii, numel( listRunsForTrain ), listRunsForTrain{ii}(1));
    for iRunTrain = 2:numel(listRunsForTrain{ii}), fprintf(', %d', listRunsForTrain{ii}(iRunTrain)); end
    fprintf(' test on runs %d', listTestRun{ii}(1) );
    for iRunTest = 2:numel(listTestRun{ii}), fprintf(', %d', listTestRun{ii}(iRunTest)); end
    
    subP3Dataset = [];
    subSsvepDataset = [];
    for iS = 1:nSub
        
        fprintf('\ttreating subject %s (%d out of %d)\n', sub{iS}, iS, nSub);
        
        %% LOAD P3 DATASET
        p3File = fullfile( resDir, '02-classify-erps' ...
            , sprintf('linSvm_%dRunsForTrain', numel( listRunsForTrain{ii} )) ...
            , sprintf('subject_%s', sub{iS}) ...
            , 'Results_forLogisiticRegression.txt' ...
            );
        
        p3Dataset = dataset('File', p3File, 'Delimiter', ',');
        
        temp = p3Dataset;
        for iRunTrain = 1:numel(listRunsForTrain{ii})
            fieldName = sprintf('trainingRun_%d', iRunTrain);
            temp = temp( ismember(temp.(fieldName), listRunsForTrain{ii}(iRunTrain)), : );
        end
        
        temp2 = [];
        for iRunTest = 1:numel(listTestRun{ii})
            temp2 = [temp2 ; temp( ismember(temp.testingRun, listTestRun{ii}(iRunTest)), : )];
        end
        subP3Dataset = [ subP3Dataset ; temp2 ];
        
        %% LOAD SSVEP DATASET
        ssvepFile = fullfile( resDir, '03-classify-ssveps' ...
            , sprintf('correctness_subject%s.txt', sub{iS}) ...
            );
        
        ssvepDataset = dataset('File', ssvepFile, 'Delimiter', ',');
        
        for iRunTest = 1:numel(listTestRun{ii})
            subSsvepDataset = [subSsvepDataset ; ssvepDataset( ismember(ssvepDataset.run, listTestRun{ii}(iRunTest)), : )];
        end
        
    end
    
    %% CHECK AND COMBINE DATASETS
    nObs = size( subSsvepDataset, 1 );
    if nObs ~= size( subP3Dataset, 1), error('sizes of dataset do not match'); end
    
    subjects = unique(subP3Dataset.subject);
    nSub = numel( subjects );
    if ~isequal( subjects, unique(subSsvepDataset.subject) ), error('subjects are different'); end
    
    runs = unique(subP3Dataset.testingRun);
    nRuns = numel( runs );
    if ~isequal( runs, unique(subSsvepDataset.run) ), error('runs are different'); end
    
    rounds = unique(subP3Dataset.roundNb);
    nRounds = numel( rounds );
    if ~isequal( rounds, unique(subSsvepDataset.roundNb) ), error('rounds are different'); end
    
    reps = unique(subP3Dataset.nAverages);
    nReps = numel( reps );
    if ~isequal( reps, unique(subSsvepDataset.nRep) ), error('rounds are different'); end

    targetFreq = unique(subP3Dataset.targetFrequency);
    nFreqs = numel( targetFreq );
    if ~isequal( targetFreq, unique(subSsvepDataset.targetFrequency) ), error('frequencies are different'); end
    
    %%
    fprintf('\tWriting down results\n');
    filename = 'train';
    for iTr = 1:numel(listRunsForTrain{ii}), filename = sprintf('%s%d', filename, listRunsForTrain{ii}(iTr)); end
    filename = sprintf('%s_test', filename);
    for iTest = 1:numel(listTestRun{ii}), filename = sprintf('%s%d', filename, listTestRun{ii}(iTest)); end
    filename = sprintf('%s.txt', filename);
    fid = fopen( fullfile( resDirOut, filename ),'wt' );
    fprintf(fid, 'subject, run, roundNb, nRep, targetFrequency, correctness_p3, correctness_ssvep, correctness\n');
    
    for iS = 1:nSub
        for iRun = 1:nRuns
            for iRound = 1:nRounds
                for iRep = 1:nReps
                    
                    dataP3 = subP3Dataset( ...
                        ismember(subP3Dataset.subject, subjects{iS}) ...
                        & ismember(subP3Dataset.testingRun, runs(iRun)) ...
                        & ismember(subP3Dataset.roundNb, rounds(iRound)) ...
                        & ismember(subP3Dataset.nAverages, reps(iRep)) ...
                        , : );
                    if size(dataP3, 1) ~= 1, error('too many observations!'); end
                    
                    dataSsvep = subSsvepDataset( ...
                        ismember(subSsvepDataset.subject, subjects{iS}) ...
                        & ismember(subSsvepDataset.run, runs(iRun)) ...
                        & ismember(subSsvepDataset.roundNb, rounds(iRound)) ...
                        & ismember(subSsvepDataset.nRep, reps(iRep)) ...
                        , :);
                    if size(dataSsvep, 1) ~= 1, error('too many observations!'); end
                    
                    if dataP3.targetFrequency ~= dataSsvep.targetFrequency, error('different target frequency!'); end

                    fprintf(fid, '%s, %d, %d, %d, %.2g, %d, %d, %d\n' ...
                        , subjects{iS} ...
                        , runs(iRun) ...
                        , rounds(iRound) ...
                        , reps(iRep) ...
                        , dataP3.targetFrequency ...
                        , dataP3.correctness ...
                        , dataSsvep.correctnessSelChan ...
                        , dataP3.correctness & dataSsvep.correctnessSelChan ...
                        );
                    
                end
            end
        end
    end
    
    fclose(fid);
end


end

File: dataAnalysisCodes/watchERP_2stim/04-icon-detection/createSymbolCorrectnessDataset.R
createSymbolCorrectnessDataset <- function(aveClass, nRunsForTrain, FS, nFoldSvm)
{
  # library(plyr)
library("reshape2")

########################################################################################################################################
########################################################################################################################################

tablename <- "d:/KULeuven/PhD/Work/Hybrid-BCI/HybBciCode/dataAnalysisCodes/watchERP_2stim/01-preprocess-plot/watchErpDataset.csv"
filelist <- read.csv(tablename, header = TRUE, sep = ",", strip.white = TRUE)
sub <- unique(filelist$subjectTag)
nSub <- length(sub)

run = unique( filelist$run )
if ( !identical(run, 1:max(run)) ){ stop("wrong run numbering") }
runsForTrain <- c(1, 2)
testRun <- run[run>max(runsForTrain)]
nRunsForTrain <- length(runsForTrain)

subsetCh <- c( "C3", "Cz", "C4", "CP5", "CP1", "CP2", "CP6", "P7", "P3", "Pz", "P4", "P8", "PO3", "PO4", "O1", "Oz", "O2" )
harmonics <- c(0, 1)

########################################################################################################################################
# P3 DATA
########################################################################################################################################


for (iS in 1:nSub){
  
  #--------------------------------------------------------------------------------------------------------------
  # Load correctness data caclulated with classifiers built on the same type of data as the test data 
  p3Folder <- file.path( "d:/KULeuven/PhD/Work/Hybrid-BCI/HybBciProcessedData/watchERP_2stim/02-classify-erps"
                       , sprintf("LinSvm_%dRunsForTrain_%dHz_%.2dcvSvm", nRunsForTrain, FS, nFoldSvm)
                       , sprintf("subject_%s", sub[iS]) )
  if (aveClass == 0){ textfile  <- "Results_forLogisiticRegression.txt" 
  } else{ textfile  <- sprintf("Results_forLogisiticRegression_%.2dAveClassifier.txt", aveClass) }
  p3file  <- file.path(p3Folder, textfile)
  p3Dataset_iS <- read.csv(p3file, header = TRUE, sep = ",", strip.white = TRUE)
  
  # Factorise
  p3Dataset_iS$foldInd          <- as.factor(p3Dataset_iS$foldInd)
  p3Dataset_iS$testingRun       <- as.factor(p3Dataset_iS$testingRun)
  p3Dataset_iS$roundNb          <- as.factor(p3Dataset_iS$roundNb)
  p3Dataset_iS$targetFrequency  <- as.factor(p3Dataset_iS$targetFrequency)
  p3Dataset_iS$nAverages        <- as.factor(p3Dataset_iS$nAverages)
  for (iTr in 1:nRunsForTrain){
    p3Dataset_iS[ , sprintf("trainingRun_%d", iTr)] <- as.factor(p3Dataset_iS[ , sprintf("trainingRun_%d", iTr)])
  }
  
  # select training run(s)
  for (iTr in 1:nRunsForTrain){
    factorname <- sprintf("trainingRun_%d", iTr)
    factorvalue <- runsForTrain[iTr]
    p3Dataset_iS <- p3Dataset_iS[ p3Dataset_iS[ , factorname] == factorvalue,  ]
  }
  
  # select test runs
  p3Dataset_iS <- p3Dataset_iS[ p3Dataset_iS$testingRun %in% testRun, ]
  
  
  # some cleaning
  p3Dataset_iS$run <- droplevels( p3Dataset_iS$testingRun )
  p3Dataset_iS$nRep <- p3Dataset_iS$nAverages    
  p3Dataset_iS <- p3Dataset_iS[c("subject", "run", "roundNb", "nRep", "targetFrequency", "correctness")]
  
  # concatenate subjects' data
  if (iS == 1) { p3Dataset <- p3Dataset_iS }
  else { p3Dataset <- rbind(p3Dataset, p3Dataset_iS) }   
}

########################################################################################################################################
# SSVEP DATA
########################################################################################################################################

for (iS in 1:nSub){
  
  # Load data
  filename  <- sprintf("d:/KULeuven/PhD/Work/Hybrid-BCI/HybBciProcessedData/watchERP_2stim/03-classify-ssveps/snrs_subjectS%d.txt", iS)
  snrDataiS   <- read.csv(filename, header = TRUE, sep = ",", strip.white = TRUE)  
  
  # Factorise
  snrDataiS$run               <- as.factor(snrDataiS$run)
  snrDataiS$roundNb           <- as.factor(snrDataiS$roundNb)
  snrDataiS$nRep              <- as.factor(snrDataiS$nRep)
  snrDataiS$targetFrequency   <- as.factor(snrDataiS$targetFrequency)
  snrDataiS$watchedFrequency  <- as.factor(snrDataiS$watchedFrequency)
  snrDataiS$harmonic          <- as.factor(snrDataiS$harmonic)
  
  # select runs, channels and harmonics
  snrDataiS <- snrDataiS[ snrDataiS$run %in% testRun & snrDataiS$harmonic %in% harmonics, ]
  if (!identical(subsetCh, "all")){ snrDataiS <- snrDataiS[ snrDataiS$channel %in% subsetCh, ] }
  snrDataiS$channel <- droplevels( snrDataiS$channel )
  snrDataiS$harmonic <- droplevels( snrDataiS$harmonic )
  snrDataiS$run <- droplevels( snrDataiS$run )
  
  # select the factors to consider
  varList <- c( "subject", "run","roundNb", "nRep", "targetFrequency", "watchedFrequency", "snr" )
  snrDataiS <- snrDataiS[ , (names(snrDataiS) %in% varList)]
  
  # average snrs over channels and harmonics
  varList <- c( "subject", "run","roundNb", "nRep", "targetFrequency", "watchedFrequency" )
  snrDataiS <- ddply(snrDataiS, varList, summarize, snr = mean(snr))
  
  # generate correctness dataset from snr values
  varList <- c( "subject", "run","roundNb", "nRep" )
  accDataiS <- ddply( snrDataiS, varList, summarize, 
                      targetFrequency = unique(targetFrequency), 
                      correctness = ( unique(targetFrequency) == watchedFrequency[which.max(snr)] ) * 1 )
  
 
  if (iS == 1) { ssvepDataset <- accDataiS }
  else { ssvepDataset <- rbind(ssvepDataset, accDataiS) }
  
}

########################################################################################################################################
# DATASET COMBINING P3, SSVEP AND SYMBOL CORRECTNESS
########################################################################################################################################

p3Dataset$type    <- as.factor("p3")
ssvepDataset$type <- as.factor("ssvep")
temp <- rbind(p3Dataset, ssvepDataset)

temp <- dcast(temp, subject + run + roundNb + nRep + targetFrequency ~ type, value.var="correctness")
temp$symbol <- (temp$p3 & temp$ssvep)*1

accDataset <- melt( temp
               , id.vars=c("subject", "run", "roundNb", "nRep", "targetFrequency")
               , measure.vars=c("p3", "ssvep", "symbol")
               , variable.name="type"
               , value.name="correctness")

# check <- test2[ test2$type=="ssvep", ]
# check$type <- droplevels( check$type )
# check <- check[ , c("subject", "run", "roundNb", "nRep", "targetFrequency", "type", "correctness")]
# check2 <- ssvepDataset[ , c("subject", "run", "roundNb", "nRep", "targetFrequency", "type", "correctness")]
# check <- check[with(check, order(subject, run, roundNb, nRep)), ]
# check2 <- check2[with(check2, order(subject, run, roundNb, nRep)), ]
# rownames(check) <- NULL
# rownames(check2) <- NULL
# identical(check, check2)

}